Sources/MongoSwift/BSON/Document.swift

```diff
@@ -177,26 +195,58 @@ public struct Document: ExpressibleByDictionaryLiteral, ExpressibleByArrayLitera
      *  ```
      */
     public subscript(key: String) -> BsonValue? {
-        get {
-            return DocumentIterator(forDocument: self, advancedTo: key)?.currentValue
-        }
-
+        get { return DocumentIterator(forDocument: self, advancedTo: key)?.currentValue }
         set(newValue) {
-            self.copyStorageIfRequired()
+            do {
+                try self.setValue(forKey: key, to: newValue)
+            } catch {
+                preconditionFailure("Failed to set the value for key \(key) to \(newValue ?? "nil"): \(error)")
+            }
+        }
+    }
 
-            guard let value = newValue else {
-                if !bson_append_null(self.data, key, Int32(key.count)) {
-                    preconditionFailure("Failed to set the value for key \(key) to null")
+    /// Sets key to newValue. if checkForKey=false, the key/value pair will be appended without checking for the key's presence first.
+    private mutating func setValue(forKey key: String, to newValue: BsonValue?, checkForKey: Bool = true) throws {
+        // if the key already exists in the `Document`, we need to replace it
+        if checkForKey, let existingType = DocumentIterator(forDocument: self, advancedTo: key)?.currentType {
+
+            let newBsonType = newValue?.bsonType ?? .null
+            let sameTypes = newBsonType == existingType
+
+            // if the new type is the same and it's a type with no custom data, no-op
+            if sameTypes && [.null, .undefined, .minKey, .maxKey].contains(newBsonType) { return }
+
+            // if the new type is the same and it's a fixed length type, we can overwrite
+            if let ov = newValue as? Overwritable, ov.bsonType == existingType {
+                self.copyStorageIfRequired()
+                // make a new iterator referencing our new storage. we already know the key is present
+                /// so initialization will succeed and ! is safe.
+                try DocumentIterator(forDocument: self, advancedTo: key)!.overwriteCurrentValue(with: ov)
+
+            // otherwise, we just create a new document and replace this key
+            } else {
+                var newSelf = Document()
```

Comment:
please make a note that this should eventually use `bson_copy_to_excluding` for performance reasons